scripts/backport.py

```diff
@@ -0,0 +1,258 @@
+
+    pulls = pygithub_commit.get_pulls()
+    if not pulls:
+        print(
+            f"{commit_sha[:9]}\terror\tlinked pr not found\t{commit_sha[:9]}\t{commit_title}"
+        )
+        continue
+
+    pull = pulls[0]
+    issue_number = get_referenced_issue(pull.number)
+    if not issue_number:
+        print(
+            f"{commit_sha[:9]}\tskip\tdoes not close an issue\t{pull.number}\t{pull.title}"
+        )
+        continue
+
+    pr_labels = {l.name for l in pull.labels}
+    if "no-backport" in pr_labels:
+        print(
+            f"{commit_sha[:9]}\tskip\tpull request marked as 'no-backport'\t{pull.number}\t{pull.title}"
+        )
+        continue
+
+    issue = source_repo.get_issue(number=issue_number)
+    issue_labels = {l.name for l in issue.labels}
+
+    if "bug" not in issue_labels:
+        print(
+            f"{commit_sha[:9]}\tskip\tissue not marked as 'bug'\t{issue.number}\t{issue.title}"
+        )
+        continue
+
+    if "no-backport" in issue_labels:
+        print(
+            f"{commit_sha[:9]}\tskip\tissue marked as 'no-backport'\t{issue.number}\t{issue.title}"
+        )
+        continue
+
+    print(f"{commit_sha[:9]}\tbackport\t\t{commit_sha}\t{commit_title}")
+
+    # Remember the PR and the corresponding resulting commit in main.
+    if pull.number not in prs_to_backport:
+        prs_to_backport[pull.number] = pull
+        pr_commits[pull.number] = [pygithub_commit]
+    else:
+        # We're traversing the history backwards, and want to have the list of
+        # commits in forward order.
+        pr_commits[pull.number].insert(0, pygithub_commit)
+
+
+# Now, go over the list of PRs that we have collected, and try to backport
+# each of them.
+for original_pr in prs_to_backport.values():
+    backport_branch = f"backport/{backport_target}/{original_pr.number}"
+
+    # If there is already a backport branch for this PR, just skip it.
+    if (
+        subprocess.run(
+            f"git rev-parse {target_remote}/{backport_branch} > /dev/null 2>&1",
+            shell=True,
+        ).returncode
+        == 0
+    ):
+        print(
+            f'Backport branch {backport_branch} for PR #{original_pr.number}: "{original_pr.title}" already exists. Skipping.'
+        )
+        continue
+
+    # Try to cherry-pick the commits.
+    subprocess.run(
+        f"git checkout --quiet --detach {source_remote}/{backport_target} > /dev/null",
+        shell=True,
+    ).check_returncode()
+
+    commit_list = " ".join([commit.sha for commit in pr_commits[original_pr.number]])
+
+    if subprocess.run(
+        # f"git cherry-pick -m 1 -x {commit_list}",
+        f"git cherry-pick -m 1 -x {commit_list} > /dev/null 2>&1",
+        shell=True,
+    ).returncode:
+        subprocess.run(
+            f"git cherry-pick --abort > /dev/null", shell=True
+        ).check_returncode()
+        print(
+            f'Conflict while backporting pull request #{original_pr.number}: "{original_pr.title}"'
```

Comment:
Should also report it to the original PR/issue. Here we need to be careful not to spam it with comment, either check for comment existence or apply a label like 'auto-backport-failed'. I think the second way might be preferrable because it's easy to filter on it, to review what we might need to backport manually.

Reply:
Added a tag for now.